### PR TITLE
#DIARIZATION SUPPORT

### PR DESCRIPTION
**Pull request by leonrch**

### CODE REVIEW

File: config/host.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+// host.js
+//module.exports.host = 'stream.watsonplatform.net';
+module.exports.host = 'stream-s.watsonplatform.net';
```
Comment (nfriedly) [inline config/host.js]: I know that you need this for your copy of the demo right now, but this should be set to the public URL before I merge it.

Ideally, I'd like to see this file removed and not even imply that the host is configurable, since it just adds extra overhead for most users.

File: manifest.yml [hunk 2]
```diff
@@ -4,7 +4,7 @@ declared-services:
     label: speech_to_text
     plan: standard
 applications:
-- name: speech-to-text-demo
+- name: speech-to-text-demo-lr
```
Comment (nfriedly) [inline manifest.yml]: This change also shouldn't be included in the PR

File: src/handlemicrophone.js [hunk 3]
```diff
@@ -54,10 +63,18 @@ exports.handleMicrophone = function(token, model, mic, callback) {
...
   };
+  
+  var keywords = display.getKeywordsToSearch();
+  if(keywords.length > 0) {
+    var keywords_threshold = 0.01;
+    options.message.keywords_threshold = keywords_threshold;
+    options.message.keywords = keywords;
+  }
+  var speaker_labels = $('li.speakersTab').is(':visible');
+  options.message.speaker_labels = speaker_labels;
```
Comment (nfriedly) [inline src/handlemicrophone.js]: Shouldn't this check the value of the checkbox rather than the visibility of the tab?

File: src/views/displaymetadata.js [hunk 4]
```diff
@@ -820,7 +836,41 @@ function onTimer() {
...
+    var token = tokenPerSpeaker.token;
+    var speaker = tokenPerSpeaker.speaker;
+    
+    if(speaker != currentSpeaker && speaker != -1) {
+      var colorClass = printf('speakerColor_%d', speaker);
+      speakers += printf("<div></div><div class='speakerInfo %s'>Speaker %d:</div>", colorClass, speaker);
+      currentSpeaker = speaker;
+    }
+    
+    if(token != '%HESITATION' && token != '~NS' && token != '~VP') {
```
Comment (nfriedly) [inline src/views/displaymetadata.js]: What do `~NS` and `~VP` mean?

**Comment by germanattanasio** [comment]
Once the changes are reviewed make sure you squash this @nfriedly

File: app.js [hunk 5]
```diff
@@ -28,6 +28,16 @@ require('dotenv').load({silent: true});
 
 // Bootstrap application settings
 require('./config/express')(app);
+var host = require('./config/host').host;
+var url = require('util').format('https://%s/speech-to-text/api', host);
+
+// production
```
Comment (germanattanasio) [inline app.js]: remove the stating/ production comment and the commented code

File: app.js [hunk 6]
```diff
@@ -37,9 +47,9 @@ app.get('/js/index.js', expressBrowserify('src/index.js', {
 // For local development, replace username and password
 var config = extend({
   version: 'v1',
-  url: 'https://stream.watsonplatform.net/speech-to-text/api',
-  username: process.env.STT_USERNAME || '<username>',
-  password: process.env.STT_PASSWORD || '<password>'
+  url: url,
```
Comment (germanattanasio) [inline app.js]: Let's keep this as it was

File: config/host.js (hunk 1 above)
Comment (germanattanasio) [inline config/host.js]: we don't need a file for this

File: manifest.yml (hunk 2 above)
Comment (germanattanasio) [inline manifest.yml]: remove this change

File: src/views/displaymetadata.js [hunk 7]
```diff
@@ -820,7 +836,41 @@ function onTimer() {
...
+      speakers += token + ' ';
+    }
+  }
+
+  console.log('*** SPEAKERS ***', speakers);
```
Comment (germanattanasio) [inline src/views/displaymetadata.js]: remove console.log

**Comment by germanattanasio** [comment]
@Mikemosca can you take a look at https://speech-to-text-demo-lr.mybluemix.net/ once you have some time?

File: src/handlefileupload.js [hunk 8]
```diff
@@ -22,24 +22,34 @@ var initSocket = require('./socket').initSocket;
...
-
+  
   $.subscribe('progress', function(evt, data) {
     console.log('progress: ', data);
   });
 
   console.log('contentType', contentType);
 
-  var baseString = '';
+  var result = {};
```
Comment (nfriedly) [inline src/handlefileupload.js]: Not a blocker, but I think this should be renamed. I was getting confused later in the code thinking that `result` was the JSON object received from the service. Maybe `state` or `output`?

File: src/views/displaymetadata.js [hunk 9]
```diff
@@ -863,20 +912,72 @@ exports.showResult = function(msg, baseString, model) {
...
+
+      if(item) {
+        if(item.final) {
+          result.speakers += diarization;
+          if($('.nav-tabs .active').text() == 'Speakers')  {
+            $('#resultsSpeakers').html(result.speakers);
+          }
+          console.log('size of tokensPerSpeaker BEFORE removeOldTokensPerSpeaker', Object.keys(tokensPerSpeaker).length);
+          removeOldTokensPerSpeaker(item.from);
+          console.log('size of tokensPerSpeaker.length AFTER removeOldTokensPerSpeaker', Object.keys(tokensPerSpeaker).length);
```
Comment (nfriedly) [inline src/views/displaymetadata.js]: These console.log's should probably be commented or removed

File: src/views/displaymetadata.js (hunk 7 above)
Comment (leonrch) [inline src/views/displaymetadata.js]: Hi German,
There are 3 three source files in the project, in which host value is used: app.js, models.js, and socket.js. My intention was to have only one place (I decided to use a separate file host.js) to define host value. Otherwise, it is so easy to forget to change it **all three** places. What is your idea about how to define host only once? Please be aware I do not mind to remove host.js. suggest

**Comment by nfriedly** [comment]
Oh, hah, I didn't realize @germanattanasio was reviewing at the same time I was. It looks like pretty similar comments.

File: config/host.js (hunk 1 above)
Comment (leonrch) [inline config/host.js]: Ok. I see your point and will remove host.js

File: src/views/displaymetadata.js (hunk 4 above)
Comment (leonrch) [inline src/views/displaymetadata.js]: They mean the same thing as %HESITATION but for different LMs (e.g. Spanish, Japanese)

File: src/handlemicrophone.js (hunk 3 above)
Comment (leonrch) [inline src/handlemicrophone.js]: Agree, checking the value of the checkbox is preferable. On the other hand, li.speakersTab is visible iff the checkbox is checked.
